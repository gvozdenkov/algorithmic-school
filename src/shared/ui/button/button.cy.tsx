import { Button } from './button';

describe('button', () => {
  var label = 'Save';
  var testClass = 'test-button';

  it('should render button with label "Save" and trigger onClick', () => {
    cy.mount(<Button text={label} data-test={testClass} onClick={cy.stub().as('click')} />);
    cy.contains(label);
    cy.getBySel(testClass).click();
    cy.get('@click').should('be.called');
  });

  it('should render disabled button', () => {
    cy.mount(<Button text={label} data-test={testClass} disabled />);
    cy.contains(label);
    cy.getBySel(testClass).should('be.disabled');
  });

  it('should render disabled button in loading state with hidden label', () => {
    cy.mount(<Button text={label} data-test={testClass} isLoader />);
    cy.contains(label);
    cy.get('svg').should('be.visible');
    cy.getBySel(testClass).should('be.disabled');
  });
});
