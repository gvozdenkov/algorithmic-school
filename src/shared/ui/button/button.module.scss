@use '#styles/media-bp-mixins.scss' as *;

.button {
  position: relative;

  display: flex;
  align-items: center;
  justify-content: center;

  min-width: 130px;
  min-height: 60px;
  padding: 18px 24px;

  color: var(--text-color-primary);

  background-color: var(--bg-button-default);
  border: none;
  border: 1px solid var(--color-border-button-default);
  border-radius: var(--common-border-radius);
  outline: none;

  transition: var(--common-transition);

  &:disabled {
    cursor: not-allowed;
    color: var(--color-text-button-disabled);
    background-color: var(--bg-button-disabled);
    border-color: var(--color-border-button-disabled);
  }

  &:hover:enabled {
    cursor: pointer;
    background-color: var(--bg-button-hover);
    border-color: var(--color-border-button-hover);
  }

  &:focus:enabled {
    background-color: var(--bg-button-active);
  }

  &:active:enabled {
    cursor: pointer;
    background-color: var(--bg-button-active);
    border-color: var(--color-border-button-active);
    fill: var(--text-color-secondary);
  }
}

.big {
  width: 362px;
}

.small {
  width: 175px;
}

.loader:disabled {
  background-color: var(--bg-button-default);
}

.loader_icon {
  position: absolute;
  top: 50%;
  left: 50%;
  translate: -50% -50%;

  animation: rotation 1s linear infinite normal forwards running;
}

.hidden-text {
  color: transparent;
}

@keyframes rotation {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}

@include mobile {
  .button {
    min-height: 44px;
    padding: 8px 12px;
  }
}
