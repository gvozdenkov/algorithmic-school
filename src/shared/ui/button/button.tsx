import { HTMLProps } from 'react';
import clsx from 'clsx';

import loaderIcon from '../../../images/icons/loader.svg';
import { AscendingIcon } from '../icons/ascending-icon';
import { DescendingIcon } from '../icons/descending-icon';
import { SortDirection } from '#shared/types';

import styles from './button.module.scss';

type ButtonProps = HTMLProps<HTMLButtonElement> & {
  text?: string;
  type?: 'button' | 'submit' | 'reset';
  sorting?: SortDirection;
  linkedList?: 'small' | 'big';
  isLoader?: boolean;
  extraClass?: string;
};

export const Button = ({
  text,
  extraClass = '',
  type = 'button',
  isLoader = false,
  sorting,
  linkedList,
  disabled,
  ...rest
}: ButtonProps) => {
  const fill = isLoader ? 'transparent' : '#e4e4e4';
  const currentIcon =
    sorting === 'asc' ? <AscendingIcon fill={fill} /> : <DescendingIcon fill={fill} />;

  return (
    <button
      className={clsx(
        'text_type_button',
        styles.button,
        { [styles.linkedList]: linkedList },
        { [styles.loader]: isLoader },
        { [extraClass]: !!extraClass },
      )}
      type={type}
      disabled={isLoader || disabled}
      {...rest}>
      {sorting && currentIcon}
      <p className={clsx('text', { 'ml-5': sorting }, { [styles.hiddenText]: isLoader })}>{text}</p>

      {isLoader && (
        <img className={styles.loader_icon} src={loaderIcon} height={'50%'} alt='Загрузка.' />
      )}
    </button>
  );
};
