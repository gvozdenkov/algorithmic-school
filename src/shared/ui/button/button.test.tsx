import { create } from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Button } from './button';

const buttonText = 'Button';
describe('Button ui', () => {
  it('with text', () => {
    const tree = create(<Button text={buttonText} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('without text', () => {
    const tree = create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('disabled', () => {
    const tree = create(<Button text={buttonText} disabled />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('loading state', () => {
    const tree = create(<Button text={buttonText} isLoader />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('actions', () => {
  it('onClick', () => {
    let newButtonText = 'new';
    render(<Button text={buttonText} onClick={() => (newButtonText = buttonText)} />);

    fireEvent.click(screen.getByText(buttonText));
    expect(buttonText).toEqual(newButtonText);
  });
});
