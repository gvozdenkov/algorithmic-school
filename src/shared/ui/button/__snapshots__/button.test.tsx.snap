// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button ui disabled 1`] = `
<button
  className="text_type_button button"
  disabled={true}
  type="button"
>
  <p
    className="text"
  >
    Button
  </p>
</button>
`;

exports[`Button ui loading state 1`] = `
<button
  className="text_type_button button loader"
  disabled={true}
  type="button"
>
  <p
    className="text hiddenText"
  >
    Button
  </p>
  <img
    alt="Loading"
    className="loader_icon"
    height="50%"
    src="loader.svg"
  />
</button>
`;

exports[`Button ui with text 1`] = `
<button
  className="text_type_button button"
  type="button"
>
  <p
    className="text"
  >
    Button
  </p>
</button>
`;

exports[`Button ui without text 1`] = `
<button
  className="text_type_button button"
  type="button"
>
  <p
    className="text"
  />
</button>
`;
